src/scenic/core/regions.py

```diff
@@ -281,883 +306,2816 @@ def __hash__(self):
+            triangles.extend(triangulatePolygon(polygon))
+        assert len(triangles) > 0, self.polygons
+        self.trianglesAndBounds = tuple((tri, tri.bounds) for tri in triangles)
+        areas = (triangle.area for triangle in triangles)
+        self.cumulativeTriangleAreas = tuple(itertools.accumulate(areas))
+
+    @cached_property
+    def footprint(self):
+        return PolygonalFootprintRegion(self.polygons)
+
+    def containsPoint(self, point):
+        return self.footprint.containsPoint(point)
+
+    def containsObject(self, obj):
+        return self.footprint.containsObject(obj)
+
+    def uniformPointInner(self):
+        triangle, bounds = random.choices(
+            self.trianglesAndBounds,
+            cum_weights=self.cumulativeTriangleAreas)[0]
+        minx, miny, maxx, maxy = bounds
+        # TODO improve?
+        while True:
+            x, y = random.uniform(minx, maxx), random.uniform(miny, maxy)
+            if triangle.intersects(shapely.geometry.Point(x, y)):
+                return self.orient(Vector(x, y, self.z))
+
+    def intersects(self, other, triedReversed=False):
+        if isinstance(other, PolygonalRegion):
+            if self.z != other.z:
+                return False
+
+        poly = toPolygon(other)
+        if poly is not None:
+            intersection = self.polygons & poly
+            return not intersection.is_empty
+        return super().intersects(other, triedReversed)
+
+    def intersect(self, other, triedReversed=False):
+        # If one of the regions isn't fixed fall back on default behavior
+        if needsSampling(self) or needsSampling(other):
+            return super().intersect(other, triedReversed)
+
+        if isinstance(other, PolygonalRegion):
+            if self.z != other.z:
+                return nowhere
+
+        poly = toPolygon(other)
+        orientation = other.orientation if self.orientation is None else self.orientation
+        if poly is not None:
+            intersection = self.polygons & poly
+            if isinstance(intersection, shapely.geometry.GeometryCollection):
+                if intersection.area > 0:
+                    poly_geoms = (shapely.geometry.Polygon, shapely.geometry.MultiPolygon)
+                    geoms = [geom for geom in intersection.geoms if isinstance(geom, poly_geoms)]
+                elif intersection.length > 0:
+                    line_geoms = (shapely.geometry.LineString, shapely.geometry.MultiLineString)
+                    geoms = [geom for geom in intersection.geoms if isinstance(geom, line_geoms)]
+                intersection = shapely.ops.unary_union(geoms)
+            orientation = orientationFor(self, other, triedReversed)
+            return regionFromShapelyObject(intersection, orientation=orientation)
+        return super().intersect(other, triedReversed)
+
+    def union(self, other, triedReversed=False, buf=0):
+        if isinstance(other, PolygonalRegion):
```

Comment:
Need the "needs sampling" checks here (although really they all need to be `isLazy` I think).